pkg/controller/backup_queue_controller.go

```diff
@@ -0,0 +1,368 @@
+/*
+Copyright the Velero contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"slices"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	corev1api "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	velerov1api "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
+	"github.com/vmware-tanzu/velero/pkg/constant"
+	"github.com/vmware-tanzu/velero/pkg/util/collections"
+	"github.com/vmware-tanzu/velero/pkg/util/kube"
+)
+
+// backupQueueReconciler reconciles a Backup object
+type backupQueueReconciler struct {
+	client.Client
+	Scheme            *runtime.Scheme
+	logger            logrus.FieldLogger
+	concurrentBackups int
+	frequency         time.Duration
+}
+
+const (
+	defaultQueuedBackupRecheckFrequency = time.Minute
```

Comment:
Do we need to make it configurable?
E.g., for small backups, the backup time may be less than 1 min

Reply:
We could make it configurable, but it may not be necessary. This one-minute recheck is really just to ensure that nothing falls through the cracks. The controller is also configured to re-reconcile when there's a new backup created and when a backup leaves InProgress, meaning when that less-than-one-min backup completes, it immediately requeues and the next one starts. I've observed that in my testing -- I created 18 backups all at once, allowing 3 concurrent backups, each one took about 10 seconds, and every time one finished, the next begun almost immediately.